### ISSUE TITLE
Add functional/integration testing with Bitcoin Core in RegTest mode

### ISSUE THREAD

**Issue by msgilligan** [bitcoinj/bitcoinj#2517]
I believe we can improve our testing by adding functional tests that have **bitcoinj** communicate with **Bitcoin Core**. This probably would be implemented by creating a new Gradle subproject (perhaps named `integration-regtest`) that requires **Bitcoin Core** -- and perhaps even automatically provisions it via [TestContainers](https://www.testcontainers.org). Whether this test module is run at build time would be configurable, of course.

The `cj-bitcoinj-spock` module in **ConsensusJ** does this and runs the tests on [GitHub Actions](https://github.com/ConsensusJ/consensusj/actions/workflows/regtest.yml).  The project https://github.com/theborakompanioni/bitcoin-spring-boot-starter seems to be using TestContainers and I think using TestContainers may make things simpler more flexible than the approach currently used in ConsensusJ.

One specific example of where this would help is it should allow us to test the `ForwardingService` example that we use in our "Getting Started" documents.

**Comment by theborakompanioni** [comment]
@msgilligan Is there anything I can do to help out?

**Comment by msgilligan** [comment]
> @msgilligan Is there anything I can do to help out?

Thanks for asking!!

I actually made an attempt at adding a `bitcoinj-integration-regtest` Gradle subproject. Of course, I used the ConsensusJ `BitcoinExtendedClient` JSON-RPC client. Unfortunately, I discovered this causes a circular dependency: the `cj-btc-jsonrpc` JAR depends upon the `bitcoinj-core` JAR, but the integration tests are running on an unreleased version of `bitcoinj-core`.

There are (at least) two solutions to this problem:
A. Run the bitcoinj integration regTests in another repository that uses a CI build of bitcoinj.  (We could do this in the ConsensusJ repository, for example)
B. Contribute the `cj-btc-jsonrpc` submodule to bitcoinj and make it a sibling of  `bitcoinj-integration-regtest`.

It seems like the next logical step would be to pursue (A) by creating a bitcoinj integration regtest module in the Conensusj/consensusj repo.

You could help out by:
1. Creating a PR for ConsensusJ that creates this module
2. Writing some GitHub (or GitLab, maybe) CI configuration that triggers the integration tests whenever the bitcoinj repo is modified. (I've done this sort of thing with Jenkins, but never with GitHub or GitLab)
3. Add TestContainers support to the existing ConsensusJ regTests and/or the work in (1).

Ultimately I would like to do (B), but @schildbach didn't seem overly enthusiastic about the idea when I presented it to him. Now that I tagged him, he'll probably comment here. However, if we implement (A), I think he'll have something more tangible to consider. Also, if we implement (A) in ConsensusJ, we could also contribute the regTest module to bitcoinj as a separate repo.

**Comment by schildbach** [comment]
I'm not "overly enthusiastic" because all those side projects need to be maintained somehow, and past has shown it can deteriorate quickly. And I feel we're still lacking on the "basics".

That said, I'm not against it either. If it can be shown that there is real value, please go for it!

**Comment by msgilligan** [comment]
Fair enough. I think it has real value both for testing what we already have and for a variety of application use cases.  I also think (or at least hope) that it will bring more usage and contributions to bitcoinj.

@theborakompanioni let's go for it!

**Comment by theborakompanioni** [comment]
> The project https://github.com/theborakompanioni/bitcoin-spring-boot-starter seems to be using TestContainers and I think using TestContainers may make things simpler and more flexible than the approach currently used in ConsensusJ.

Currently I am trying to move away from testcontainers and start regtest environments with [`spring-boot-docker-compose`](https://docs.spring.io/spring-boot/docs/3.1.0/reference/html/features.html#features.docker-compose). Would this be a viable solution for a new integration test package? I can can start a draft in a PR and add some basic tests, maybe you can then decide if it is applicable for consensusj/bitcoinj/other projects. Is this approach okay for you?

**Comment by msgilligan** [comment]
Why are you moving away from Testcontainers?

How tied to Spring is `spring-boot-docker-compose`?  I am also (eventually) interested in integration testing with TESTNET, MAINNET, and SIGNET chains. I suppose that could be supported, too.

It sounds like it could be viable, I'd be interested in looking at the draft PR...

**Comment by theborakompanioni** [comment]
> Why are you moving away from Testcontainers?

For more complex test setups, it became cumbersome to wire Testcontainers together (core, electrum, lightning nodes, etc.). A simple `docker-compose.yml` files with all services is more streamlined and easier to comprehend. Although, if it is just a single Bitcoin Core service, there is nothing to be said against the use of Testcontainers.

> How tied to Spring is `spring-boot-docker-compose`? I am also (eventually) interested in integration testing with TESTNET, MAINNET, and SIGNET chains. I suppose that could be supported, too.

Tightly coupled to spring-boot, I understand that this might not be what you want. However, if it is just for integration test purposes, maybe it's not that bad after all. But as mentioned above, for a single service, a Testcontainer should be sufficient.

Regarding other networks than `regtest`: I suppose this would have to be a persistent and permantently running instance, which will incur costs. I would not know how to set up something like this in a reliable test suite for bitcoinj.

> It sounds like it could be viable, I'd be interested in looking at the draft PR...

:+1:
